clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,976 @@
+
+        private final ReadWriteLock lock = new ReentrantReadWriteLock();
+        private final Condition subscriptionLoaded = lock.writeLock().newCondition();
+        private final Condition terminalPushInProgress = lock.writeLock().newCondition();
+        /*
+         Initial state should be subscription needed which should allow issuing first telemetry
+         request of get telemetry subscription.
+        */
+        private ClientTelemetryState state = ClientTelemetryState.SUBSCRIPTION_NEEDED;
+
+        private ClientTelemetrySubscription subscription;
+
+        /*
+         Last time a telemetry request was made. Initialized to 0 to indicate that no request has
+         been made yet. Telemetry requests, get or post, should always be made after the push interval
+         time has elapsed.
+        */
+        private long lastRequestMs;
+        /*
+         Interval between telemetry requests in milliseconds. Initialized to 0 to indicate that the
+         interval has not yet been computed. The first get request will be immediately triggered as
+         soon as the client is ready.
+        */
+        private int intervalMs;
+        /*
+         Whether the client telemetry sender is enabled or not. Initialized to true to indicate that
+         the client telemetry sender is enabled. This is used to disable the client telemetry sender
+         when the client receives unrecoverable error from broker.
+        */
+        private boolean enabled;
+
+        private DefaultClientTelemetrySender() {
+            enabled = true;
+        }
+
+        @Override
+        public long timeToNextUpdate(long requestTimeoutMs) {
+            final long now = time.milliseconds();
+            final ClientTelemetryState localState;
+            final long localLastRequestMs;
+            final int localIntervalMs;
+
+            lock.readLock().lock();
+            try {
+                if (!enabled) {
+                    return Integer.MAX_VALUE;
+                }
+
+                localState = state;
+                localLastRequestMs = lastRequestMs;
+                localIntervalMs = intervalMs;
+            } finally {
+                lock.readLock().unlock();
+            }
+
+            final long timeMs;
+            final String apiName;
+            final String msg;
+
+            switch (localState) {
+                case SUBSCRIPTION_IN_PROGRESS:
+                case PUSH_IN_PROGRESS:
+                    /*
+                     We have a network request in progress. We record the time of the request
+                     submission, so wait that amount of the time PLUS the requestTimeout that
+                     is provided.
+                    */
+                    apiName = (localState == ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS) ? ApiKeys.GET_TELEMETRY_SUBSCRIPTIONS.name : ApiKeys.PUSH_TELEMETRY.name;
+                    timeMs = requestTimeoutMs;
```

Comment:
don't we need to wait for the network request to be completed first as specified above? A side question, when timeToNextUpdate() is invoked, I assume the the caller will wait for that amount of time and try to push again, is that the use case? 

if requestTimeoutMs the client side configuration?